src/Illuminate/Database/Query/Builder.php

```diff
@@ -3749,12 +3749,18 @@ public function updateFrom(array $values)
      * Insert or update a record matching the attributes, and fill it with values.
      *
      * @param  array  $attributes
-     * @param  array  $values
+     * @param  \Closure|array  $values
      * @return bool
      */
-    public function updateOrInsert(array $attributes, array $values = [])
+    public function updateOrInsert(array $attributes, $values = [])
```

Comment:
I think we can keep the typing in this function.

Reply:
> I think we can keep the parameter typing in this function.

Although the function itself will return an array, the argument being passed will still be a function so it will throw a `TypeError` based on that.

Reply:
Use a union type, then? 

```php
public function updateOrInsert(array $attributes, callable|array $values = [])
```

Reply:
> Use a union type, then?
> 
> ```
> public function updateOrInsert(array $attributes, callable|array $values = [])
> ```

Definitely. I'll amend my PR to update it.

I'm currently working in an old project running on PHP 7.4, so am unable to use union types. I just made the amends locally and then made the PR with what worked for me in that setup without accounting for the PHP 8 support in Laravel 11.